### PR TITLE
[tlul] xbar data blanking

### PR DESCRIPTION
**Pull request by tjaychen**
- xbar portion fix to #11036
- Technically this should be applied to m1 as well for reads

Signed-off-by: Timothy Chen <timothytim@google.com>

### CODE REVIEW

**Comment by tjaychen** [comment]
This is not meant to be merged yet since #11036 needs either TC approval or wider discussion in general.
This just shows one possible solution to the presented problem.

File: hw/ip/tlul/rtl/tlul_socket_1n.sv
Comment (tjaychen) [inline hw/ip/tlul/rtl/tlul_socket_1n.sv:168]: the blanking is intentionally done before the FIFOs.

Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:168]: this LGTM

File: hw/ip/tlul/rtl/tlul_socket_1n.sv
Comment (msfschaffner) [inline hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: Regarding the choice of these values, we may have to discuss this a bit more...

Specifically, for the `CLKMGR.JITTER_ENABLE` register where this problem has initially been observed, it just occurred to me that blanking this to a valid `MuBi4False` value would still make it possible to disable the jittery clock by faulting the CSR write enable.

Would it maybe make more sense to set this to a value that is not a valid MUBI encoding? 

Or should we set the bus-integrity to an invalid value?
I guess that would require us to check the data integrity at a later point in the CSR node since right now integrity is not checked at the level of individual registers...

**Review by msfschaffner — COMMENTED** [review]
Thanks for this draft, @tjaychen!

File: hw/ip/tlul/rtl/tlul_socket_1n.sv
Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: o yeah good point. Instead of a mubi value, this is be an invalid mubi value.
And hopefully downstream we have good strict / false checks that pushes into the right bucket.

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: yeah i thought about the itnegrity also..but like you said..if the glitch is theoretically done on the write enable of the final flop... there is almost no check we could do at that point.

Actually here i'm already supplying bad integrity, but if that completely bypasses the checks because the fault point is just later and relying on the bus broadcast... it skips right around.

Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: > o yeah good point. Instead of a mubi value, this should be an invalid mubi value. And hopefully downstream we have good strict / false checks that pushes into the right bucket.

exactly, that sounds good!

Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: > yeah i thought about the itnegrity also..but like you said..if the glitch is theoretically done on the write enable of the final flop... there is almost no check we could do at that point.
> 
> Actually here i'm already supplying bad integrity, but if that completely bypasses the checks because the fault point is just later and relying on the bus broadcast... it skips right around.

Ah thanks for claryfing - I overlooked that.
Yeah so for the later fault point... the solutions (for expanding coverage) that I can currently think of are along the lines of
1) store ECC alongside each CSR and check it concurrently (like the buffer regs in OTP). but since not each CSR is 32bit, this approach would need some sort of re-encoding as well.
2) distribute ECC to all CSRs but don't flop it. instead, gate the bits with the individual CSR WE and gather them again with a large mux or OR reduction and check.
But these solutions seem to be rather heavy...

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: I do think solution number 1 is something we have to do long term honestly.
Mostly because shadow registers are so expensive we have to provide an intermediate option.

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: option 2 we could reasonably do now as an enhancement without too much impact right? what do you think?

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: > option 2 we could reasonably do now as an enhancement without too much impact right? what do you think? Although this might add an ECC encoder to every WE.. is that what you're thinking?

Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: yeah option two is certainly more lightweight than option 1 and it would increase coverage.
I guess impact is limited since it is all autogen, it is limited to the 7bits ECC, and it would not change ECC behavior/encoding...

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: how do you envision this working though? if we distribute ECC to each register, does that mean each one will separately decode ECC to see if it's matching with the data, and then qualify the write enable? 

I'm not sure what was originally discussed, but i had thought the attack would be downstream of all of that directly at the WE of the flop?

Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: The attack was downstream on the WE of the flop, that is correct.

At the moment, I believe we only check ECC on valid incoming bus transactions, independently of the individual flop WE signals - so that would not be able to detect scenarios where a flop WE is glitched downstream.

There are several ways we could improve protection against such attacks. I.e. like you say route ECC to flops and check it at each flop when the corresponding WE is asserted. What I sketched above in 2 is similar, but would use a reduction / mux first before feeding ECC back into one single checker (instead of using multiple ECC checkers).

Another way would be to just gather all flop WEs and trigger an additional ECC check whenever any of the WEs is asserted (that way we sort of move this check further downstream). Or we could just add a general crosscheck of the WE signals and check whether there are any spurious activations. WDYT?

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: i think i'm not completely following the reduction method. If it's not done after the fact, wouldn't you still have the issue that the WE is attacked directly, and whatever is on the bus gets loaded in?  My main thought is that anything used to qualify the WE line could potentially be defeated if the attack is close enough to the flop. 

The second one is interesting to me. So I gather we would sample all the WEs and in the cycle after check if the data written and integrity agreed? (which it obviously will not).

Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: I guess the methods I have mentioned so far assumed that whenever a flop WE is forced / glitched to 1, that would also propagate into any checks that are qualified by that WE (so it's trying to move the check downstream of WE).

So for the WE gathering idea, my initial thought was to do that combinationally in the same cycle.
But you make a good point in that we could increase coverage even more if we delayed this check into the next cycle so that we can check actual register contents. That would become something like a hybrid between bus-integrity and the shadow reg (since essentially we're not shadowing each reg, but rather keep one shared copy around to compare with in the next cycle).

Note however that we may have to do a verbatim data comparison in that case, since not all CSRs latch the entire 32bit (so re-calculating ECC based on CSR contents in the next cycle won't work for these).

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: do you want to first check that assumption of WE glitch?
The reason I was assuming "next" cycle is mostly because of an assumption that it's attacked around the rising edge of the clock.  So there may or may not be enough time to even evaluate in the current cycle or capture on the next. 

I also wasn't quite thinking comparing the CSR contents... but more..."there was a write last cycle, did the data / integrity actually agree"? I think you were suggesting something very similar on the current cycle, that could work too. 

If I understand, what you're basically saying is that if we ever see an assertion of WE (regardless of whether there is a valid transaction at the front end), we check with the integrity?

Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: Hm yes the timing around the clock edge is a fair point. Where you thinking about latching the individual WE bits alongside the CSRs and using them to trigger a comparison in the next cycle? I think that could work too.

Yes I was basically looking for a solution that would monitor all WE bits (regardless of whether the incoming transaction was valid), and trigger some sort of crosscheck based on that. The check itself could be an ECC check, a data comparison, or even just a plain comparison with the incoming write enable (which in case of spurious WE assertions should fail, too).

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: sounds good. let me think a bit about this. I think in general this should be good, but off the top of my head i'm not sure what's the right approach for the async registers.  Since those are technically in a different domain.

Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: We can probably do something similar and just monitor the WE strobes in the other clock domain (we probably have to sync the main write-valid signal over to be able to do that check). There should not be any spurious WEs there either.

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: okay i took a quick look, i kind of forgot after the tlul_adapter there is a pile of address decoding combo logic.
So let's take the raw "we" and compare it against the combined "we" nearest the actual field registers. 

My current thinking is simply to OR all the decoded ones together and see any of them show a 1 vs the original. 

for the async ones i will likely capture the "we" across the clock boundary and compare that way. 

Back to the original one, if we don't want the blanking value to `MuBiFalse`, then it would probably need to be '0' or all '1's?
For mubi4, it will still have a hamming delta of 2 to both polar values.

Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: Ok that approach sounds good to me!
And for the async one, that is what I have been thinking as well.

Yes, I believe either all-zero or all-ones should be fine.
Which one to pick probably depends a bit on which value we believe does less "damage".
However, that might be a bit of a toss-up, depending on the IPs we're looking at...

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: yeah i was thinking about that..there's not a good way to go imo.
if we pick 0's, then anything that's opt-in is at risk.
if we pick 1's, then anything that's opt-out is at risk. 

My gut feeling is that we have more opt-in's, but don't know for sure.

Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:154]: Same here - I have the feeling all-ones could be slightly safer... (but again that is just a feeling).

File: hw/ip/tlul/rtl/tlul_socket_1n.sv
Comment (tjaychen) [inline hw/ip/tlul/rtl/tlul_socket_1n.sv:144]: @msfschaffner have a look at this part.
I'll take a look at the reg portion now, just wanted to get this part out of the way first.

File: hw/ip/tlul/rtl/tlul_socket_1n.sv
Comment (msfschaffner) [inline hw/ip/tlul/rtl/tlul_socket_1n.sv:157]: Should this comment be placed above the `localparam logic [top_pkg::TL_DW-1:0] BlankedAData = {top_pkg::TL_DW{1'b1}};` instead? Having down here led me to think that this is about integrity itself within `a_user`, which is not really the case.

Also, since the blanked A data is all-ones, you could technically assign all-ones to the ECC as well, since the all-ones 39bit codeword is not valid per design.

File: hw/ip/tlul/rtl/tlul_socket_1n.sv
Comment (msfschaffner) [inline hw/ip/tlul/rtl/tlul_socket_1n.sv:150]: Would inverting the existing integrity bits be cheaper than a full-re-encode?

**Comment by cdgori** [comment]
so far this makes sense.  I don't have a strong intuition for opt-in vs opt-out, but I think I can see your argument.

File: hw/ip/tlul/rtl/tlul_socket_1n.sv
Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:144]: LGTM with two questions regarding the invalid ECC functions.

**Review by msfschaffner — APPROVED** [review]

File: hw/ip/tlul/rtl/tlul_socket_1n.sv
Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:157]: yeah i can move the comment.
I thought about that, but i didn't want to rely on that and wanted something explicitly wrong always. 
So even if we swap how we do things underneath, it remains incorrect.

File: hw/ip/tlul/rtl/tlul_socket_1n.sv
Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:150]: so the only reason i didn't do that because i wasn't completely sure that a full inversion would always be wrong :). Could we land into one of those 1/N cases? That's partially why, just like the data below, i went for something explicitly incorrect.

File: hw/ip/tlul/rtl/tlul_socket_1n.sv
Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:157]: That sounds good to me as well - just wanted to point that out.

File: hw/ip/tlul/rtl/tlul_socket_1n.sv
Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:150]: Oh I see what you are getting at... there are likely some corner cases where that would not work, that is correct. 

I guess the question is whether adding another ECC encoder to all sockets is ok complexity wise?

Also, do we care about these corner cases if we can guarantee that the data integrity is incorrect?

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:150]: so specifically to the original vector that was mentioned... neither the data integrity nor the command integrity would offer any kind of protection.  Since the attack is all downstream of that. 

in terms of complexity...i don't think it's too bad because there really arent that many socket1n's. If we really do care, the right thing to do would be to fix the blanked bus output (similar to data), then the "bad" integrity would become a constant.  I was thinking about that too, but just got lazy.

In terms of whether it helps, yes I think it helps a little.  Because ultimately, IF the attack is not right at the flop, but somewhere upstream-ish, it just gives more points to have to get right.

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:150]: maybe we should go for the fixed blank output, and thus fixed "bad" command integrity.

Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:150]: > so specifically to the original vector that was mentioned... neither the data integrity nor the command integrity would offer any kind of protection. Since the attack is all downstream of that.

yeah that is correct - unless we also add something like a "late" ECC check either at the register, or in the cycle after (as initially brainstormed further above). 

> in terms of complexity...i don't think it's too bad because there really arent that many socket1n's. If we really do care, the right thing to do would be to fix the blanked bus output (similar to data), then the "bad" integrity would become a constant. I was thinking about that too, but just got lazy.

in terms of complexity the fixed blanked but output is probably somewhat preferable... but I am fine with the current solution if you think the fixed blanked output would be difficult to add.

> In terms of whether it helps, yes I think it helps a little. Because ultimately, IF the attack is not right at the flop, but somewhere upstream-ish, it just gives more points to have to get right.

yeah it makes things harder to attack, that is true.

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:150]: i'll go with the fixed approach, you're right that we probably shouldn't needlessly add logic here when we don't HAVE to.

Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:150]: sounds good!

**Review by eunchan — APPROVED** [review]

**Review by weicaiyang — APPROVED** [review]
LGTM, let me add some assertions to check this

File: hw/ip/tlul/rtl/tlul_socket_1n.sv
Comment (tjaychen) [inline hw/ip/tlul/rtl/tlul_socket_1n.sv:144]: @eunchan 
@msfschaffner 

can you guys have a look? Especially to make sure i didn't do something wrong since I rearranged this code a bit.

File: hw/ip/tlul/rtl/tlul_pkg.sv
Comment (msfschaffner) [inline hw/ip/tlul/rtl/tlul_pkg.sv:82]: This might cause a lint error, hence I would recommend to use `{top_pkg::TL_DW{1'b1}}` instead.

File: hw/ip/tlul/rtl/tlul_socket_1n.sv
Comment (msfschaffner) [inline hw/ip/tlul/rtl/tlul_socket_1n.sv:164]: nit: resopnses -> responses

File: hw/ip/tlul/rtl/tlul_socket_1n.sv
Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:144]: This LGTM!

**Review by msfschaffner — APPROVED** [review]

File: hw/ip/tlul/rtl/tlul_pkg.sv
Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_pkg.sv:82]: i didn't see one. This only caused a lint error for me at when placing it into a parameter, which is why i previously did the concat.

File: hw/ip/tlul/rtl/tlul_socket_1n.sv
Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:144]: @vogelpi 
@a-will 

hmmm... could this coding (and more importantly below) confuse vivado and end up synthesizing something much larger? the latest fpga error is that there are no more LUTs. But the previous version of this PR, which actually wasn't even a constant was fine..

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:144]: it might also be the extra muxing? it's a bit surprising... since i would have assumed previously having a live ecc encoding would be worse. I'll experiment a bit and see what happens.

Comment (a-will) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:144]: Hm... we're apparently looking at an extra ~400 LUTs and an extra ~2000 registers, according to the raw synthesis log. Is this something that might be multiplying a seemingly modest number by many users?

I wish I had the synthesis checkpoint from CI so I could dive deeper. The CW305 build is a bit more annoying to kick off locally.

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:144]: wait, sorry where are you seeing an extra 2000 registers? 
since i'm using `always_comb` here, shouldn't this be all logic only?

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:144]: o just realized you might be referring to the results not the code..

Comment (a-will) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:144]: Oops! It's actually 2000 registers less for this build vs. a recent one on master. I should probably find the build from your base branch...

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:144]: yea i'm running that locally now

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:144]: poking around with @a-will , it really looks like it's just the added logic.
The original ECC encoded approach ended up being smaller since (i'm guessing) the command payload has a lot of hardwring (address, source id etc), so the xor's tree ends up quite small. The final muxing is also small since it's mostly on the integrity bits.

Whereas if I muxed the entire bus, there are many more bits involved (even though one side of the mux is completely hardcoded). So i'll go back to the original approach.

@vogelpi 
As a side note, should i take out `usbdev` from top_englishbreakfast? I might also add some parameters to disable memory scrambling to free up some more room.

Comment (a-will) [reply hw/ip/tlul/rtl/tlul_socket_1n.sv:144]: I think most of these signals were largely passed from one FIFO to another (i.e. no combinational logic between the state elements for most of them). Consequently, the new muxes added a new LUT for each output, and that multiplied across a large number of bits and each of dozens of instances. Because the LUT is the smallest discrete unit in an FPGA, that made this change especially painful.

In the ASIC, the area used might be considerably less, though the multiplicative factor is still there.

File: hw/ip/tlul/rtl/tlul_pkg.sv
Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_pkg.sv]: ok interesting. the reason I brought this up is that we generally discourage the use of unsized literals (other than `'0`)... 
see [here](https://github.com/lowRISC/style-guides/blob/master/VerilogCodingStyle.md#always-be-explicit-about-the-widths-of-number-literals).

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_pkg.sv]: yeah now i'm wondering if i was just blind yesterday. But in any case, i had to change this around again and i went back to the concat.

File: hw/ip/tlul/rtl/tlul_pkg.sv
Comment (msfschaffner) [inline hw/ip/tlul/rtl/tlul_pkg.sv:78]: Can this potentially be replaced with `DataWhenError` introduced in this PR: https://github.com/lowRISC/opentitan/pull/11324?

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_pkg.sv:78]: heh, i was actually going to suggest in the other PR to be replaced with this one :)
Did we make the other one all 1's? i can't remember now.

Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_pkg.sv:78]: Yes we set it to all-ones - although the parameter name differs a bit since we differentiate between `DataWhenError` and `DataWhenInstrError` on the `d_data` path.

I guess we need to align this once either of the two PRs has been merged.

Comment (tjaychen) [reply hw/ip/tlul/rtl/tlul_pkg.sv:78]: i can align, since i need to make one more push anyways to squash the commits.
do you want to merge the other first?

Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_pkg.sv:78]: ok sounds good. it should be almost there - only waiting for the usual FPGA CI and Verilator tests to complete.

File: hw/ip/tlul/rtl/tlul_pkg.sv
Comment (rswarbrick) [reply hw/ip/tlul/rtl/tlul_pkg.sv]: I don't think that the text in the style guide actually applies to `'1`: the "all zeros" and "all ones" unsized literals are genuinely different from e.g. "123". We might decide that we want to tighten up the text there, but at the moment, I don't think it has anything to say about using `'1` like this.

Comment (a-will) [reply hw/ip/tlul/rtl/tlul_pkg.sv]: Oh, thanks for giving those a name, since I apparently started using it without knowing what it was. I have some fixes to make in usbdev...

**Comment by tjaychen** [comment]
merging this part for now and will begin on part 2.

File: hw/ip/tlul/rtl/tlul_pkg.sv
Comment (msfschaffner) [reply hw/ip/tlul/rtl/tlul_pkg.sv]: @rswarbrick that guidance was derived from an internal style guide, and in that one `'1` was explicitly called out to be avoided since it can be confused for a `1` (as opposed to auto-extended `1111...1111`). The only unsized literal that should be used is `'0`.

**Pull request merged**
